# MSMC Tools

This repository contains helper scripts and utilities for [msmc](http://github.com/stschiff/msmc) and [msmc2](http://github.com/stschiff/msmc). Most scripts have a little built in help, which can be invoked with the option -h.

### generate_multihetsep.py
This script is the main workhorse to generate input files for msmc. Here is a short synopsis:

    ./generate_multihetsep.py --mask=covered_sites_sample1_chr1.bed.txt.gz \
                              --mask=covered_sites_sample1_chr1.bed.txt.gz \
                              --mask=mappability_mask_chr1.bed.txt.gz \
                              sample1_chr1.vcf.gz sample2_chr1.vcf.gz

In this example, you generate multihetsep files for two diploid individuals. For each individual you need a pair of sample-specific input files: One single-sample/single-chromosome VCF file, and one mask file in bed format, which gives the regions on the chromosome on which the genome of that individual was covered sufficiently. This pair of mask- and vcf files can for example be generated by one my caller scripts (`bamCaller.py` or `cgCaller.py`).

In addition to the mask- and vcf file for each individual, you should use one additional mask per chromosome, the mappability mask, which gives all regions on the chromosome, on which short sequencing reads can be uniquely mapped. For humans (GRCh37), you can download these masks from my ftp site, see the readme for msmc.

You can also apply masks in a negative way, using the option `--negative_mask`, which declares which regions should be _excluded_ from the analysis, instead of _included_, as with `--mask`. This is useful for example if you have an admixed sample and have a bed file which gives all regions of non-native ancestry that you would like to mask out. Any mask file, either by this option or by `--mask` can be given gzipped, using the standard file ending `.gz`.

### bamCaller.py

* `--legend_file`: If you aim to phase your data against a reference panel, e.g. from 1000 Genomes (see section below about Phasing), you need your VCF to not only contain the variant sites of the sample, but also the genotypes at additional sites at which the panel is genotyped. This option takes a gzipped file of a format that is used in the IMPUTE and SHAPEIT reference panels. It is a simple tab-separated tabular file format with one header line which gets ignored. The only important columns for this purpose are: 1. the chromosome; 2. the position; 3. the reference allele; 4. the alternative allele; 5. the type of the variant, only sites of type `SNP` are considered here.


### cgCaller.py

### run_shapeit.sh
### getStats.d
### multihetsep_bootstrap.py
